lib/dynamoid/dumping.rb

```diff
@@ -297,10 +297,14 @@ def process(value)
       end
     end
 
-    # string -> string
+    # string -> StringIO
     class BinaryDumper < Base
       def process(value)
-        Base64.strict_encode64(value)
+        if value.is_a?(StringIO) || value.is_a?(IO)
+          value
+        else
+          StringIO.new(value)
+        end
```

Comment:
I suppose that now new values of a binary field will be stored in DynamoDB as `binary` but all existing values are stored as `string`. It might be a bit surprising for end users.

The common approach is to steadily migrate users from the current behaviour to the new one. We can introduce a new field's option to either keep old behaviour (by default, so there is no a breaking change in a minor release) or to switch to the new one (for new fields or new tables or in new projects).

Example is a `boolean` type that wasn't supported by DynamoDB initially so Dynamoid's `boolean` fields were stored as `f` and `t`. Later DynamoDB added support of the boolean type and a new option `store_as_native_boolean` was added:

```ruby
field :active, :boolean, store_as_native_boolean: false
```

Later in the next major release such migration options can be removed and we can propose users to  custom types to keep old storage formats.

So I propose to add a new field's option `store_as_native_binary` that is by default false. And check it at casting, dumping and undumping.

Reply:
I agree. It could cause issues for integrators even though they might not have noticed that the binary values aren't stored with binary type in DynamoDB.

I've added the `store_as_native_binary` option that should be a safe work-around, but there's still a risk with this change to not get noticed with the new major release since there will not be a breaking code change.

This concerns me along with the lack of alignment with DynamoDB's single table design which is why I've started exploring `Aws::DynamoDB::Client` directly for my project.

Let me know if you want me to do any other changes to this PR, or I'll leave it to you for how you want to get it out.

Reply:
The only way to prevent breaking changes being unnoticed that I see is to add deprecation warnings in the last minor release before the major one.

Reply:
Could also require the `store_as_native_binary` option be set, otherwise raise an error.

Reply:
Regarding the single-table design - it seems to me it's out of scope of Dynamoid. Or at least of the initial scope to implement the Active Record pattern.

I am not opposed to including the new functionality into the gem related to this "pattern". But I haven't found any common and widely excepted "interface" or API. So I postponed this.

There is an issue #568 but all I heard there were either some generic thoughts or requesting somebody's project specific features. So there was nothing I can work with.